torch_xla/distributed/xla_backend.py

```diff
@@ -0,0 +1,289 @@
+        if last_shape != t.shape:
+          raise ValueError(f"Input {i+1}'s shape is different from input {i}: "
+                           f"{t.shape} vs {last_shape}")
+        last_shape = t.shape
+      input_tensor = torch.cat(input_tensors)
+      reduce_type = self._get_reduce_type(opts.reduceOp)
+      groups = self._mesh
+      shard_count = len(groups[0]) if groups else self.size()
+      result = xm.reduce_scatter(
+          reduce_type,
+          input_tensor,
+          scatter_dim=0,
+          shard_count=shard_count,
+          scale=1,
+          groups=groups)
+
+      output_tensor.set_(result)
+
+    return WorkXla(output_tensors)
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L2683
+  def barrier(self, opts):
+    return WorkXla()
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L1417
+  # `reduce` is not needed by DeepSpeed for now.
+  def reduce(self, *args):
+    raise NotImplementedError
+
+  def allgather_coalesced(self, *args):
+    raise NotImplementedError
+
+  def allreduce_coalesced(self, *args):
+    raise NotImplementedError
+
+  def alltoall(self, *args):
+    raise NotImplementedError
+
+  def alltoall_base(self, *args):
+    raise NotImplementedError
+
+  def gather(self, *args):
+    raise NotImplementedError
+
+  def scatter(self, *args):
+    raise NotImplementedError
+
+  def send(self, *args):
+    raise NotImplementedError
+
+  def recv(self, *args):
+    raise NotImplementedError
+
+  def recv_anysource(self, *args):
+    raise NotImplementedError
+
+  def monitored_barrier(self, *args):
+    raise NotImplementedError
+
+  def Options(self, *args):
+    raise NotImplementedError
+
+
+class WorkXla(Work):
+
+  def __init__(self, cc_tensors=None):
+    '''
+        Args:
+            cc_tensors (List[torch.Tensor]): List of `torch.Tensor`s that
+            have collective communication ops pending.
+            For each Tensor `t` in the list, `t.device` must be an `xla`
+            device.
+        '''
+    super().__init__()
+    self.cc_tensors = cc_tensors
+
+  def wait(self):
+    if self.cc_tensors is not None:
+      if distutils.util.strtobool(
+          os.environ.get('XLA_BACKEND_BLOCKING_CC_OPS', 'False')):
+        logging.info("XLA Backend: Waiting for tensor CC op...")
+        torch_xla._XLAC._xla_sync_multi(self.cc_tensors, devices=[], wait=True)
+      else:
+        logging.info("XLA Backend: Skipping tensor CC op wait.")
+    else:
+      if distutils.util.strtobool(
+          os.environ.get('XLA_BACKEND_BLOCKING_BARRIER', 'False')):
+        logging.info("XLA Backend: Non-tensor wait...")
+        torch_xla._XLAC._xla_step_marker(
+            torch_xla._XLAC._xla_get_default_device(), devices=[], wait=True)
+      else:
+        logging.info("XLA Backend: Skipping non-tensor wait.")
```

Comment:
@JackCaoG need your help to review this function. With CUDA tensors, the expected behavior of calling wait is that the current CUDA stream will wait for the collective comm op creating this tensor.

Reply:
what I am not sure about is the behavior of the following code:

```python
work1 = all_reduce(tensor)
work2 = all_reduce(tensor)
# with CUDA, this would only wait for the 1st allreduce. What about XLA?
work1.wait()
```

Reply:
Here is my thinking
1. Pytorch/XLA don't really need `wait`, every communication op should just build the graph and move on. Pytorch/XLA should return a tensor with `all_reduce` HLO and the correct shape/type but nothing should be evaluated until the `mark_step` is called.
2. If we really have to use `wait`, it should be a `mark_step` which will execute all of the pending graph instead of only evaluating the sub-graph. Breaking the graph into sub-graph will have the speed implication.

Reply:
on a second thought, I am strongly leaning toward making `wait` a `no-op`. This will help user to generate the larger graph which helps the performance. This would also save user from the danger of recompilations(due to `wait` inserted in random places).

Reply:
Does XLA support streams? If yes, what if application code called `work.wait()` in a user stream X and rely on that to block subsequent ops in that stream? If `wait` becomes no-op, such ordering won't be preserved?

Reply:
XLA has stream concept at the runtime level but we are deprecating that, that didn't happen until we actually execute the graph through. Could you give me an example of how this `wait` is usually used? From xla's perspective, we are just building graph and execute in the end.

Reply:
```python
x = torch.ones(2, 2)
y = torch.ones(3, 3)
work = all_reduce(tensor, async_op=True)
new_stream = torch.cuda.Stream()
with torch.cuda.stream(new_stream):
    work.wait()
    xx = x + 1  # this will run after all_reduce
yy = y + 1  #  this will run concurrently with all_reduce
```

Above is a toy example, but we do use streams in FSDP and DDP to enable 1) concurrent computation and communication 2) and use `wait` to make sure that optimizers won't consume grad values before they are ready. 

Would I be correct that XLA will get the graph and use that graph to automatically figure out which comm/comp can be overlapped?

Reply:
I am aligned with Jack's thinking. The current implementation is that, by default, wait is a no-op. Only if user specifies `XLA_BACKEND_BLOCKING_CC_OPS` or `XLA_BACKEND_BLOCKING_BARRIER` will we really wait.

Reply:
>  In a sense that there is only one stream per execution. And PyTorch/XLA will ensure that there is at most one execution happens at a given time.

In this case, are we still going to have overlap for communication and computation?

Reply:
Xla Compiler will make sure all of the following computation blocked by the communication op. There won't be overlap for communication and computation.